DiscordRPC/dllmain.cpp

```diff
@@ -0,0 +1,227 @@
+// dllmain.cpp : Defines the entry point for the DLL application.
+#include "pch.h"
+#include "dllmain.h"
+#include <array>
+#include <cassert>
+#include <csignal>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <thread>
+#include <vector>
+#include <windows.h>
+#include <chrono>
+
+#include "library/discord.h"
+
+#if defined(_WIN32)
+#pragma pack(push, 1)
+
+
+
+struct BitmapImageHeader
+{
+	uint32_t const structSize{sizeof(BitmapImageHeader)};
+	int32_t width{0};
+	int32_t height{0};
+	uint16_t const planes{1};
+	uint16_t const bpp{32};
+	uint32_t const pad0{0};
+	uint32_t const pad1{0};
+	uint32_t const hres{2835};
+	uint32_t const vres{2835};
+	uint32_t const pad4{0};
+	uint32_t const pad5{0};
+
+	BitmapImageHeader& operator=(BitmapImageHeader const&) = delete;
+};
+
+struct BitmapFileHeader
+{
+	uint8_t const magic0{'B'};
+	uint8_t const magic1{'M'};
+	uint32_t size{0};
+	uint32_t const pad{0};
+	uint32_t const offset{sizeof(BitmapFileHeader) + sizeof(BitmapImageHeader)};
+
+	BitmapFileHeader& operator=(BitmapFileHeader const&) = delete;
+};
+#pragma pack(pop)
+#endif
+
+BOOL APIENTRY DllMain(HMODULE hModule, DWORD ul_reason_for_call, LPVOID lpReserved)
+{
+	switch (ul_reason_for_call)
+	{
+	case DLL_PROCESS_ATTACH:
+	case DLL_THREAD_ATTACH:
+	case DLL_THREAD_DETACH:
+	case DLL_PROCESS_DETACH:
+		break;
+	}
+	return TRUE;
+}
+
+struct DiscordState
+{
+	discord::User currentUser;
+
+	std::unique_ptr<discord::Core> core;
+};
+
+namespace
+{
+volatile bool interrupted{false};
+}
+
+#define EXPORT __declspec(dllexport)
+
+#include "../NorthstarDedicatedTest/state.h"
+
+GameState* gameStatePTR = 0;
+
+extern "C" EXPORT void initializePlugin(GameState* gameStatePTR_external)
+{
+	gameStatePTR = gameStatePTR_external;
+	std::thread discord(main, 0, (char**)0);
+	discord.detach();
+}
+
+DiscordState state{};
+bool wasInGame;
+
+int main(int, char**)
+{
+
+	SetEnvironmentVariable(L"DISCORD_INSTANCE_ID", L"0");
```

Comment:
https://discord.com/developers/docs/game-sdk/sdk-starter-guide

> If you test with this, make sure to remove this code before pushing a production build. It will interfere with the way that Discord launches games for users.